src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -182,33 +181,29 @@ public function handle(AddProductToOrderCommand $command)
 
             Hook::exec('actionOrderEdited', ['order' => $order]);
 
-            $newCartRules = $cart->getCartRules();
+            $cartRules = $cart->getCartRules();
 
-            sort($oldCartRules);
-            sort($newCartRules);
+            foreach ($cartRules as $cartRuleInArray) {
+                if (empty($order->getCartRules())) {
+                    $this->createNewOrUpdateExistingOrderCartRule(new OrderCartRule(), $cartRuleInArray, $order, $invoice);
 
-            // Serialize permits to diff multi dimensional array
-            $result = array_diff(
-                array_map('serialize', $newCartRules),
-                array_map('serialize', $oldCartRules)
-            );
-            $result = array_map('unserialize', $result);
-
-            foreach ($result as $cartRule) {
-                // Create OrderCartRule
-                $rule = new CartRule($cartRule['id_cart_rule']);
-                $values = [
-                    'tax_incl' => $rule->getContextualValue(true),
-                    'tax_excl' => $rule->getContextualValue(false),
-                ];
-                $orderCartRule = new OrderCartRule();
-                $orderCartRule->id_order = $order->id;
-                $orderCartRule->id_cart_rule = $cartRule['id_cart_rule'];
-                $orderCartRule->id_order_invoice = !empty($invoice->id) ? $invoice->id : 0;
-                $orderCartRule->name = $cartRule['name'];
-                $orderCartRule->value = $values['tax_incl'];
-                $orderCartRule->value_tax_excl = $values['tax_excl'];
-                $orderCartRule->add();
+                    continue;
+                }
+
+                foreach ($order->getCartRules() as $orderCartRule) {
+                    if ($orderCartRule['id_cart_rule'] !== $cartRuleInArray['id_cart_rule']) {
```

Comment:
This double loop feels weird, I have the feeling it could create too many OrderCartRule, especially if they are not in the same order. Example:

```php
$cartRuleInArray = [
  ['id_cart_rule' => 1],
  ['id_cart_rule' => 2],
];

$order->getCartRules() = [
  ['id_cart_rule' => 2],
  ['id_cart_rule' => 1],
];
```

With the current implementation it will believe there is no match because it doesn't loop through the whole list to find the matching id.

I trie something similar but slightly different (I also find it easier to read/understand):

```php
        foreach ($cartRules as $cartRule) {
                $rule = new CartRule($cartRule['id_cart_rule']);

                // Search order cart rule
                $orderCartRule = null;
                foreach ($order->getCartRules() as $orderCartRule) {
                    if ($orderCartRule['id_cart_rule'] === $cartRule['id_cart_rule']) {
                        $orderCartRule = new OrderCartRule($orderCartRule['id_order_cart_rule']);

                        break;
                    }
                }

                if (null === $orderCartRule) {
                    $orderCartRule = new OrderCartRule();
                }
                $orderCartRule->id_order = $order->id;
                $orderCartRule->id_cart_rule = $cartRule['id_cart_rule'];
                $orderCartRule->id_order_invoice = !empty($invoice->id) ? $invoice->id : 0;
                $orderCartRule->name = $cartRule['name'];
                $orderCartRule->value = $rule->getContextualValue(true);
                $orderCartRule->value_tax_excl = $rule->getContextualValue(false);
                $orderCartRule->save();
            }
```

Reply:
agree :+1: